File: examples/python-client/converter_client.py
#!/usr/bin/env python3

import requests
import json

class CurrencyConverterClient:
    def __init__(self, base_url="http://localhost:8080"):
        self.base_url = base_url
    
    def convert(self, amount, from_currency, to_currency):
        """Convert currency amount"""
        url = f"{self.base_url}/convert"
        data = {
            "amount": amount,
            "from": from_currency,
            "to": to_currency
        }
        
        response = requests.post(url, json=data)
        if response.status_code == 200:
            return response.json()
        else:
            raise Exception(f"Conversion failed: {response.json()}")
    
    def get_currencies(self):
        """Get supported currencies"""
        url = f"{self.base_url}/currencies"
        response = requests.get(url)
        if response.status_code == 200:
            return response.json()["currencies"]
        else:
            raise Exception(f"Failed to get currencies: {response.json()}")
    
    def set_rate(self, from_currency, to_currency, rate):
        """Set custom exchange rate"""
        url = f"{self.base_url}/rates"
        data = {
            "from": from_currency,
            "to": to_currency,
            "rate": rate
        }
        
        response = requests.post(url, json=data)
        if response.status_code != 200:
            raise Exception(f"Failed to set rate: {response.json()}")

def main():
    client = CurrencyConverterClient()
    
    print("=== Currency Converter Python Client ===")
    
    # Get supported currencies
    print("\n1. Supported currencies:")
    currencies = client.get_currencies()
    print(currencies)
    
    # Convert USD to EUR
    print("\n2. Convert $100 USD to EUR:")
    result = client.convert(100, "USD", "EUR")
    print(f"${result['originalAmount']} USD = €{result['convertedAmount']:.2f} EUR")
    print(f"Exchange rate: {result['exchangeRate']}")
    
    # Convert EUR to JPY
    print("\n3. Convert €50 EUR to JPY:")
    result = client.convert(50, "EUR", "JPY")
    print(f"€{result['originalAmount']} EUR = ¥{result['convertedAmount']:.2f} JPY")
    
    # Set custom rate and convert
    print("\n4. Set custom USD->EUR rate to 0.90:")
    client.set_rate("USD", "EUR", 0.90)
    
    print("\n5. Convert $100 USD to EUR with custom rate:")
    result = client.convert(100, "USD", "EUR")
    print(f"${result['originalAmount']} USD = €{result['convertedAmount']:.2f} EUR")
    print(f"Custom exchange rate: {result['exchangeRate']}")

if __name__ == "__main__":
    main()

File: sbets-system/pkg/client/converter_client.go
package client

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
)

type ConverterClient struct {
	baseURL string
}

type ConvertRequest struct {
	Amount float64 `json:"amount"`
	From   string  `json:"from"`
	To     string  `json:"to"`
}

type ConvertResponse struct {
	ConvertedAmount float64 `json:"convertedAmount"`
	OriginalAmount  float64 `json:"originalAmount"`
	FromCurrency    string  `json:"fromCurrency"`
	ToCurrency      string  `json:"toCurrency"`
	ExchangeRate    float64 `json:"exchangeRate"`
}

func NewConverterClient() *ConverterClient {
	baseURL := os.Getenv("CONVERTER_SERVICE_URL")
	if baseURL == "" {
		baseURL = "http://localhost:8085"
	}
	return &ConverterClient{baseURL: baseURL}
}

func (c *ConverterClient) Convert(amount float64, from, to string) (*ConvertResponse, error) {
	req := ConvertRequest{
		Amount: amount,
		From:   from,
		To:     to,
	}

	jsonData, err := json.Marshal(req)
	if err != nil {
		return nil, err
	}

	resp, err := http.Post(c.baseURL+"/convert", "application/json", bytes.NewBuffer(jsonData))
	if err != nil {
		return nil, fmt.Errorf("failed to call converter service: %v", err)
	}
	// When Convert() finishes, close the response body.
	defer resp.Body.Close()

	print("RESPONSE FROM CONVERTER:  ", resp)

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("converter service returned status %d", resp.StatusCode)
	}

	var result ConvertResponse
	// resp.Body is an open stream (important!)
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}

	return &result, nil
}
